org.eclipse.ui.genericeditor/src/org/eclipse/ui/internal/genericeditor/hover/TextHoverRegistry.java

```diff
@@ -89,32 +93,44 @@ public TextHoverRegistry(IPreferenceStore preferenceStore) {
 		Platform.getExtensionRegistry().addRegistryChangeListener(event -> outOfSync = true, EXTENSION_POINT_ID);
 	}
 
-	public List<ITextHover> getAvailableHovers(ISourceViewer sourceViewer, ITextEditor editor, Set<IContentType> contentTypes) {
+	public List<ITextHover> getAvailableHovers(ISourceViewer sourceViewer, ITextEditor editor,
+			Set<IContentType> contentTypes) {
 		if (this.outOfSync) {
 			sync();
 		}
-		return this.extensions.stream()
-				.filter(ext -> contentTypes.contains(ext.targetContentType))
+		return this.extensions.stream().filter(ext -> contentTypes.contains(ext.targetContentType))
 				.filter(ext -> ext.matches(sourceViewer, editor))
-				// don't sort in the stream as the initial structure is already sorted by isAfter/isBefore
-				.map(GenericContentTypeRelatedExtension<ITextHover>::createDelegate)
+				// don't sort in the stream as the initial structure is already sorted by
+				// isAfter/isBefore
+				.map(ext -> {
+					ITextHover delegateInstance = ext.createDelegate();
+					if (sourceViewer instanceof ISourceViewerExtension6
+							&& delegateInstance instanceof ITextViewerLifecycle) {
+						((ISourceViewerExtension6) sourceViewer)
+								.addTextViewerLifecycle((ITextViewerLifecycle) delegateInstance);
+					}
+					return delegateInstance;
```

Comment:
The registry should not be responsible for calling the addTextViewerLifecycle here.
Instead, the CompositeTextHover should implement ITextViewerLifecycle and make the install/uninstall delegate to its children; and the ISourceViewer should check whether the provided ITextHover implements ITextViewerLifcycle and call install/uninstall on it when appropriate.

Reply:
Thanks so much for your suggestion. I did that for IAutoEditingStrategy. I don't do that for ITextHover since ITextHover receives the ITextViewer.

The goal of ItextViewerLifecycle is to get the ItextViewer with a clean mean,as ITextHover receives the ITextViewer as parameter, I think it is not very helpful.